README.md

Gives polyline coordinates to set polylines in Google Map.

[![pub package](https://img.shields.io/pub/v/google_map_polyline.svg)](https://pub.dartlang.org/packages/google_map_polyline)

# google_map_polyline
Flutter plugin to retrieve coordinates (in latitude and longitude) to draw the polylines (Route) in `google_maps_flutter` package.
This package made with the inspiration of `flutter_polyline_points`.

### Getting Started

To use this plugin, add `google_map_polyline` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

### Example

1. First you need to create object using your **Google Maps API key**.
- Get an API key at [https://cloud.google.com/maps-platform/](https://cloud.google.com/maps-platform/).
	 (This plugin only requires **Directions API**.)

	```dart
		GoogleMapPolyline googleMapPolyline = 
			new  GoogleMapPolyline(apiKey:  "YOUR KEY HERE");
	```
		
2. Now you may able retrieve coordinates for the Polyline route.

- Get the Polyline coordinates using location coordinates (latitude and longitude)
	 ```dart
     	await googleMapPolyline.getCoordinatesWithLocation(
     		origin: LatLng(40.677939, -73.941755),
     		destination: LatLng(40.698432, -73.924038),
     		mode:  RouteMode.driving);	
	```


- Get the Polyline coordinates using address of the location
	```dart
		await googleMapPolyline.getPolylineCoordinatesWithAddress(
			origin:  '55 Kingston Ave, Brooklyn, NY 11213, USA',
			destination:  '8007 Cypress Ave, Glendale, NY 11385, USA',
			mode:  RouteMode.driving);
	```

### Overview
- [x] Coordinates with Location Coordinates (Latitude and Longitude)
- [x] Coordinates with Location Address
- [x] Route Modes
  	- [x] Driving
  	- [x] Walking
  	- [x] Bicycling
- [ ] Coordinates with Location Coordinates and Address (example: Origin as address and Destination as coordinates)
- [ ] Alternative routes


### Feature Requests and Issues
Please file feature requests and bugs at the  [issue tracker](https://github.com/Shark01/google_map_polyline/issues/new).


CHANGELOG.md
## [0.0.1 - 0.0.3+1]
- Initial publish

## [0.0.4]
- Updated dependencies

## [0.2.0 - 0.2.0+1]
- Minor changes

example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:google_map_polyline/google_map_polyline.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int _polylineCount = 1;
  Map<PolylineId, Polyline> _polylines = <PolylineId, Polyline>{};
  GoogleMapController _controller;

  GoogleMapPolyline _googleMapPolyline =
      new GoogleMapPolyline(apiKey: "YOUR KEY HERE");

  //Polyline patterns
  List<List<PatternItem>> patterns = <List<PatternItem>>[
    <PatternItem>[], //line
    <PatternItem>[PatternItem.dash(30.0), PatternItem.gap(20.0)], //dash
    <PatternItem>[PatternItem.dot, PatternItem.gap(10.0)], //dot
    <PatternItem>[
      //dash-dot
      PatternItem.dash(30.0),
      PatternItem.gap(20.0),
      PatternItem.dot,
      PatternItem.gap(20.0)
    ],
  ];

  LatLng _mapInitLocation = LatLng(40.683337, -73.940432);

  LatLng _originLocation = LatLng(40.677939, -73.941755);
  LatLng _destinationLocation = LatLng(40.698432, -73.924038);

  bool _loading = false;

  _onMapCreated(GoogleMapController controller) {
    setState(() {
      _controller = controller;
    });
  }

  //Get polyline with Location (latitude and longitude)
  _getPolylinesWithLocation() async {
    _setLoadingMenu(true);
    List<LatLng> _coordinates =
        await _googleMapPolyline.getCoordinatesWithLocation(
            origin: _originLocation,
            destination: _destinationLocation,
            mode: RouteMode.driving);

    setState(() {
      _polylines.clear();
    });
    _addPolyline(_coordinates);
    _setLoadingMenu(false);
  }

  //Get polyline with Address
  _getPolylinesWithAddress() async {
    _setLoadingMenu(true);
    List<LatLng> _coordinates =
        await _googleMapPolyline.getPolylineCoordinatesWithAddress(
            origin: '55 Kingston Ave, Brooklyn, NY 11213, USA',
            destination: '8007 Cypress Ave, Glendale, NY 11385, USA',
            mode: RouteMode.driving);

    setState(() {
      _polylines.clear();
    });
    _addPolyline(_coordinates);
    _setLoadingMenu(false);
  }

  _addPolyline(List<LatLng> _coordinates) {
    PolylineId id = PolylineId("poly$_polylineCount");
    Polyline polyline = Polyline(
        polylineId: id,
        patterns: patterns[0],
        color: Colors.blueAccent,
        points: _coordinates,
        width: 10,
        onTap: () {});

    setState(() {
      _polylines[id] = polyline;
      _polylineCount++;
    });
  }

  _setLoadingMenu(bool _status) {
    setState(() {
      _loading = _status;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      darkTheme: ThemeData(brightness: Brightness.dark),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Map Polyline'),
        ),
        body: Container(
          child: LayoutBuilder(
            builder: (context, cont) {
              return Column(
                children: <Widget>[
                  SizedBox(
                    width: MediaQuery.of(context).size.width,
                    height: MediaQuery.of(context).size.height - 175,
                    child: GoogleMap(
                      onMapCreated: _onMapCreated,
                      polylines: Set<Polyline>.of(_polylines.values),
                      initialCameraPosition: CameraPosition(
                        target: _mapInitLocation,
                        zoom: 15,
                      ),
                    ),
                  ),
                  Expanded(
                    child: Align(
                        alignment: Alignment.center,
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                          children: <Widget>[
                            RaisedButton(
                              child: Text('Polylines wtih Location'),
                              onPressed: _getPolylinesWithLocation,
                            ),
                            RaisedButton(
                              child: Text('Polylines wtih Address'),
                              onPressed: _getPolylinesWithAddress,
                            ),
                          ],
                        )),
                  ),
                ],
              );
            },
          ),
        ),
        floatingActionButtonLocation: FloatingActionButtonLocation.centerDocked,
        floatingActionButton: _loading
            ? Container(
                color: Colors.black.withOpacity(0.75),
                child: Center(
                  child: Text(
                    'Loading...',
                    style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
                  ),
                ),
              )
            : Container(),
      ),
      debugShowCheckedModeBanner: false,
    );
  }
}
